crates/net/network/src/fetch/mod.rs

```diff
@@ -236,24 +246,42 @@ impl StateFetcher {
         let is_error = res.is_err();
         let maybe_reputation_change = res.reputation_change_err();
 
-        let resp = self.inflight_headers_requests.remove(&peer_id);
+        let request = self.inflight_headers_requests.remove(&peer_id);
 
-        let is_likely_bad_response = resp
+        let is_likely_bad_response = request
             .as_ref()
             .map(|r| res.is_likely_bad_headers_response(&r.request))
             .unwrap_or_default();
 
-        if let Some(resp) = resp {
-            // delegate the response
-            let _ = resp.response.send(res.map(|h| (peer_id, h).into()));
-        }
-
+        let mut send_follow_up = !is_error && !is_likely_bad_response;
         if let Some(peer) = self.peers.get_mut(&peer_id) {
             // If the peer is still ready to accept new requests, we try to send a followup
             // request immediately.
-            if peer.state.on_request_finished() && !is_error && !is_likely_bad_response {
-                return self.followup_request(peer_id)
+            send_follow_up &= peer.state.on_request_finished();
+        }
+
+        if let Some(request) = request {
+            if res.as_ref().is_ok_and(|res| res.is_empty()) {
```

Comment:
this is already covered with `is_likely_bad_response` and can be reused here I believe